java-samples/src/main/java/com/google/cloud/testing/SpannerEmulator.java

```diff
@@ -0,0 +1,250 @@
+package com.google.cloud.testing;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.spanner.Database;
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.DatabaseId;
+import com.google.cloud.spanner.InstanceConfig;
+import com.google.cloud.spanner.InstanceId;
+import com.google.cloud.spanner.InstanceInfo;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.junit.rules.ExternalResource;
+
+/**
+ * Junit ClassRule to create and shut down a Cloud Spanner Emulator with a Spanner database for a
+ * unit test.
+ *
+ * <p>This rule should be used as a static {@link org.junit.ClassRule} which will be run before the
+ * JUnit test class is instantiated, and after all tests are complete.
+ *
+ * <pre>{@code
+ * @ClassRule
+ * public static SpannerEmulator emulator = SpannerEmulator.builder()
+ *     .projectId(PROJECT_ID)
+ *     .instanceId(INSTANCE_ID)
+ *     .databaseId(DATABASE_ID)
+ *     .emulatorPort(EMULATOR_PORT)
+ *     .emulatorRestPort(EMULATOR_REST_PORT)
+ *     .ddlStatements(
+ *         "CREATE TABLE table1 (key INT64, value STRING(MAX)) PRIMARY KEY(key)",
+ *         "CREATE TABLE table2 (key INT64, value STRING(MAX)) PRIMARY KEY(key)")
+ *     .build()
+ *
+ * }</pre>
+ *
+ * <p>Default values are available, so an empty database with no schema can be build using:
+ *
+ * <pre>{@code
+ * @ClassRule
+ * public static SpannerEmulator emulator = SpannerEmulator.builder().build()
+ *
+ * }</pre>
+ *
+ * <p>However to avoid port number collisions, it is recommended to specify port numbers for {@link
+ * Builder#emulatorPort(int)} and {@link Builder#emulatorRestPort(int)}
+ *
+ * <p>A {@link Spanner} instance can be returned using {@link SpannerEmulator#getSpanner()}, and a
+ * {@link DatabaseClient} using {@link SpannerEmulator#getDatabaseClient()}, so during a test,
+ * values can be read from or written to the database using {@code emulator.getDatabaseClient()}
+ *
+ * <p>Requires <a
+ * href="https://github.com/google/auto/blob/master/value/userguide/index.md">AutoValue</a> and <a
+ * href="https://github.com/google/guava">Guava</a> to be part of your build environment.
+ */
+@AutoValue
+public abstract class SpannerEmulator extends ExternalResource {
+
+  private static final int DEFAULT_EMULATOR_PORT = 29010;
+  private static final int DEFAULT_EMULATOR_REST_PORT = 29020;
+
+  static {
+    // This Test Rule uses Linux kill command, so only works on Unix based OS.
+    if (!System.getProperty("os.name").toLowerCase().contains("linux")) {
```

Comment:
I'm not sure exactly where/how this sample came into existence, but as a general note: We also start an emulator during tests for our Liquibase tests: https://github.com/cloudspannerecosystem/liquibase-spanner/blob/e97a5b3597f4084e669a88c3117e071573cde988/src/test/java/com/google/spanner/liquibase/TestHarness.java#L146

This solution uses the Docker image instead of the `gcloud` integration, and has the additional benefit that it works on both Linux and MacOS (and maybe also Windows, haven't tried it for as far as I can remember). That might be worth considering for a separate sample.